artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/server/impl/RemotingServiceImpl.java

```diff
@@ -407,15 +408,28 @@ public void stop(final boolean criticalError) throws Exception {
          conn.disconnect(criticalError);
       }
 
+      Map<Acceptor, Future<?>> acceptorFutures = new HashMap<>();
       for (Acceptor acceptor : acceptors.values()) {
          try {
-            acceptor.stop();
+            acceptorFutures.put(acceptor, acceptor.asyncStop());
          } catch (Throwable t) {
             ActiveMQServerLogger.LOGGER.errorStoppingAcceptor(acceptor.getName());
          }
       }
 
+      long acceptorTimeout = System.currentTimeMillis() + 3000;
+      for (Map.Entry<Acceptor, Future<?>> acceptorFuture : acceptorFutures.entrySet()) {
+         if (acceptorFuture.getValue() != null) {
+            try {
+               acceptorFuture.getValue().get(Math.max(0, acceptorTimeout - System.currentTimeMillis()), TimeUnit.MILLISECONDS);
```

Comment:
Why not simply call get(3000) every time here?


Can you declare the constant as a private final field... 

Also, explain why there is a timeout here with a comment

Reply:
In a scenario with 2 acceptors if the first takes 2 seconds and the second takes 4 seconds the loop takes 4 seconds.

Reply:
What happens to the acceptors that dont close in time? Are the port bindings just left hanging?

Reply:
This is actually a fix for NetworkFailureFailoverTest. So no need for a test.


(@brusdev  please mention this is fixing networkFailureFailoverTest after reapplying the commit)


NetworkFailureFailoverTest was actually added to prevent a situation where a network card gone wouldn't prevent a server from stopping.

Instead of hunging during a shutdown the server will wait on a timeout, and the process will die and the OS will take care of cleaning up resources on that failure case.


On a regular scenario these will run really fast and never take very long.

Reply:
@michaelandrepearce also I reverted a previous fix as networkFailurefailoverTest was broken, and this is fixing it.

Reply:
I assume though that test case existed already right? And was previously passing.....

So why is this needed, what broke it, and why. Very cautious in areas like this. Sorry if im sucking eggs

Reply:
@michaelandrepearce Domenico added a rule to validate if File Descriptors are temporarily leaking. The fix was to block until the acceptor was stopped.

That broke NetworkFailureFailoverTest, as the block was infinite

Now the fix is apply a timeout on the stop, which will fix both cases.

Reply:
Has a physical test been done to prove the network one still works? If i remember we found it the hard way a year or two back and it took a physcial retest to prove it, i realise the test tried to simulate but the times a phsycial test has found issue where simulated didnt.

Just i am super cautious here, as these are failure scenarious already been bitten by

Reply:
I'm not concerned as there is no semantic change if a timeout is applied.

I'm actually glad the test caught this situation right on the testsuite. It would been a regression if it wasn't for that test.


I will make a manual test before merging it. Although this test is automating how I used to test this manually (by adding and removing a network card through ifconfig)

Reply:
This has to be merged anyways.. as I know it is broken now in master.

I'm just working thorugh a better way to do these timeouts.

Reply:
im glad too :) , but you can understand my nervousness right :)